File: ValidParentheses.cpp
class Solution {
public:
	bool isValid(string s) {
		map<char, char> hm;
		stack<char> st;
		hm.insert(make_pair('(', ')'));
		hm.insert(make_pair('{', '}'));
		hm.insert(make_pair('[', ']'));

		for (char c : s) {
			map<char, char>::iterator it = hm.find(c);
			if (it != hm.end())
				st.push(c);
			else {
				if (st.empty())
					return false;
				char st_top = st.top();
				st.pop();
				it = hm.find(st_top);
				if (it->second != c)
					return false;
			}
		}
		return st.empty();
	}
};

File: ValidParentheses.java
public class Solution {
	public boolean isValid(String s) {
		HashMap<Character, Character> hm = new HashMap<>();
		Stack<Character> st = new Stack<>();
		hm.put('(', ')');
		hm.put('{', '}');
		hm.put('[', ']');

		char[] ch = s.toCharArray();

		for (char c : ch) {
			if (hm.containsKey(c))
				st.push(c);
			else if (hm.containsValue(c)) {
				if (st.empty() || hm.get(st.pop()) != c)
					return false;
			}
			else
				return false;
		}
		return st.empty();
	}
}
